freqtrade/plugins/pairlist/MarketCapPairList.py

```diff
@@ -0,0 +1,92 @@
+"""
+MarketCap PairList provider
+
+Provides dynamic pair list based on market cap
+"""
+import logging
+from typing import Any, Dict, List
+
+import requests
+from cachetools.ttl import TTLCache
+
+from freqtrade.exceptions import OperationalException
+from freqtrade.plugins.pairlist.IPairList import IPairList
+
+
+logger = logging.getLogger(__name__)
+
+
+class MarketCapPairList(IPairList):
+    def __init__(self, exchange, pairlistmanager,
+                 config: Dict[str, Any], pairlistconfig: Dict[str, Any],
+                 pairlist_pos: int) -> None:
+        super().__init__(exchange, pairlistmanager, config, pairlistconfig, pairlist_pos)
+
+        if 'number_assets' not in self._pairlistconfig:
+            raise OperationalException(
+                '`number_assets` not specified. Please check your configuration '
+                'for "pairlist.config.number_assets"')
+
+        self._stake_currency = config['stake_currency']
+        self._number_pairs = self._pairlistconfig['number_assets']
+        self._refresh_period = self._pairlistconfig.get('refresh_period', 6*60*60)
+        self._pair_cache: TTLCache = TTLCache(maxsize=1, ttl=self._refresh_period)
+        self._url = \
+            f'https://api.coingecko.com/api/v3/coins/markets?vs_currency=usd&order=market_cap_desc&per_page={self._number_pairs}&page=1&sparkline=false'   # noqa
+        if not self._exchange.exchange_has('fetchTickers'):
+            raise OperationalException(
+                'Exchange does not support dynamic whitelist. '
+                'Please edit your config and restart the bot.'
+            )
+
+    @property
+    def needstickers(self) -> bool:
+        """
+        Boolean property defining if tickers are necessary.
+        If no Pairlist requires tickers, an empty Dict is passed
+        as tickers argument to filter_pairlist
+        """
+        return True
+
+    def short_desc(self) -> str:
+        """
+        Short whitelist method description - used for startup-messages
+        """
+        return f"{self.name} - top {self._pairlistconfig['number_assets']} marketcap pairs."
+
+    def gen_pairlist(self, tickers: Dict) -> List[str]:
+        """
+        Generate the pairlist
+        :param tickers: Tickers (from exchange.get_tickers()). May be cached.
+        :return: List of pairs
+        """
+        # Generate dynamic whitelist
+        # Must always run if this pairlist is not the first in the list.
+        pairlist = self._pair_cache.get('pairlist')
+        if not pairlist:
+            resp = requests.get(self._url)
+            data = resp.json()
```

Comment:
:+1: I'll change that.

Reply:
reimplemented using pycoingecko.

The errors from pycoingecko are less specific than I'd like them to be unfortunately, but I think I've handled it in a sane way. pycoingecko retries requests in case they fail, so I haven't implemented that.